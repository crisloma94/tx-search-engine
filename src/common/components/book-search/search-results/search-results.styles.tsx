import styled, { css } from 'styled-components';

export const StyledSearchResults = styled.div(
  ({ theme: { colors, spaces } }) => css`
    position: absolute;
    background-color: ${colors.lightSecondary};
    width: 30rem;
    right: 0;
    max-height: 80vh;
    overflow: hidden;
    overflow-y: scroll;
    margin: ${spaces.small};
    padding: ${spaces.medium};

    .search-result__list {
      padding: 0;
      margin-bottom: ${spaces.large};
    }
  `,
);
