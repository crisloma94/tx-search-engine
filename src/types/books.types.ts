// Defined in DEFAULT_BOOK_FIELDS
export interface BookSearchModel {
  author_name?: string[];
  cover_i?: number;
  first_publish_year?: number;
  id_goodreads?: string[];
  id_librarything?: string[];
  isbn?: string[];
  key: string;
  oclc?: string[];
  publisher?: string[];
  title: string;
}
