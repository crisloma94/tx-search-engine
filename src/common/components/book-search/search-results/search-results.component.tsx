import { forwardRef, Ref } from 'react';
import { BookSearchModel } from '../../../../types/books.types';
import BookCard from '../../book-card/book-card.component';
import Button from '../../button/button.component';
import SearchResultsSkeleton from './search-results-skeleton/search-results-skeleton.component';
import { StyledSearchResults } from './search-results.styles';

export interface SearchResultsProps {
  results: BookSearchModel[];
  loading: boolean;
  error: string | null;
  loadMore: () => void;
}

export const SearchResults = forwardRef<HTMLDivElement, SearchResultsProps>(
  (
    { results = [], loading, error, loadMore }: SearchResultsProps,
    ref: Ref<HTMLDivElement>,
  ) => {
    return (
      <StyledSearchResults ref={ref}>
        {!!results.length && (
          <ul className="search-result__list">
            {results.map((book) => (
              <BookCard book={book} key={book.key} />
            ))}
          </ul>
        )}
        {loading && <SearchResultsSkeleton />}
        {!error && !loading && !!results.length && (
          <Button text="Load more" onClick={loadMore} />
        )}
        {!error && !loading && !results.length && (
          <p>We could not find any books</p>
        )}
        {error && <p>{error}</p>}
      </StyledSearchResults>
    );
  },
);
