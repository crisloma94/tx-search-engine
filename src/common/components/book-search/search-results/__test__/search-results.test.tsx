import { render, screen } from '@testing-library/react';
import { SearchResults, SearchResultsProps } from '../search-results.component';
import { BookSearchModel } from '../../../../../types/books.types';
import SearchResultsSkeleton from '../search-results-skeleton/search-results-skeleton.component';
import { Provider } from 'react-redux';
import Theme from '../../../../../styles/theme';
import { store } from '../../../../../core/store/store';

describe('Search results component test suite', () => {
  const defaultBook: BookSearchModel = {
    author_name: ['Author name test'],
    cover_i: 5678,
    first_publish_year: 1984,
    id_goodreads: ['1234'],
    id_librarything: ['567'],
    isbn: ['89'],
    key: '/works/OL545572W',
    oclc: ['1', '2'],
    publisher: ['Publisher example'],
    title: 'Title test',
  };

  const loadMoreMock = jest.fn();

  const getComponent = (props: SearchResultsProps) => (
    <Provider store={store}>
      <Theme>
        <SearchResults {...props} />
      </Theme>
    </Provider>
  );
  const setup = (props: SearchResultsProps) => render(getComponent(props));

  it('Should match snapshot', () => {
    const container = setup({
      results: [defaultBook],
      loading: false,
      error: null,
      loadMore: loadMoreMock,
    });
    expect(container).toMatchSnapshot();
  });

  it('Should render the loading component when loading is true', () => {
    setup({
      results: [defaultBook],
      loading: true,
      error: null,
      loadMore: loadMoreMock,
    });
    expect(<SearchResultsSkeleton />).toBeTruthy();
  });

  it('Should render the error component when error is true', () => {
    const errorText = 'There has been an error... try again';
    setup({
      results: [],
      loading: false,
      error: errorText,
      loadMore: loadMoreMock,
    });
    expect(screen.getByText(errorText)).toBeInTheDocument();
  });

  it('Should render the no results component when results is empty', () => {
    setup({ results: [], loading: false, error: '', loadMore: loadMoreMock });
    expect(screen.getByText('We could not find any books')).toBeInTheDocument();
  });

  it('Should render the load more button when results is not empty', () => {
    setup({
      results: [defaultBook],
      loading: false,
      error: '',
      loadMore: loadMoreMock,
    });
    expect(screen.getByText('Load more')).toBeInTheDocument();
  });
});
