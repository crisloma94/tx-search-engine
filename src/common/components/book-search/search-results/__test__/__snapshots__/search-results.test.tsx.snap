// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search results component test suite Should match snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="sc-dicizt fDjvmk"
      >
        <ul
          class="search-result__list"
        >
          <li
            class="sc-gTRrQi gGrCkx"
          >
            <img
              alt="Default book cover"
              class="sc-bgqQcB ccLtkH book-cover"
              src="https://covers.openlibrary.org/b/id/5678-M.jpg"
            />
            <div
              class="book-card__info"
            >
              Title test (1984)
              <p
                class="book-card__author"
              >
                Author name test
              </p>
              <p
                class="book-card__publisher"
              >
                Publisher example
              </p>
              <a
                href="https://www.amazon.es/s?k=Title test+Author name test"
                target="_blank"
              >
                Search on amazon
              </a>
            </div>
          </li>
        </ul>
        <button
          class="sc-ewnqHT ctMhSl"
          type="button"
        >
          Load more
        </button>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="sc-dicizt fDjvmk"
    >
      <ul
        class="search-result__list"
      >
        <li
          class="sc-gTRrQi gGrCkx"
        >
          <img
            alt="Default book cover"
            class="sc-bgqQcB ccLtkH book-cover"
            src="https://covers.openlibrary.org/b/id/5678-M.jpg"
          />
          <div
            class="book-card__info"
          >
            Title test (1984)
            <p
              class="book-card__author"
            >
              Author name test
            </p>
            <p
              class="book-card__publisher"
            >
              Publisher example
            </p>
            <a
              href="https://www.amazon.es/s?k=Title test+Author name test"
              target="_blank"
            >
              Search on amazon
            </a>
          </div>
        </li>
      </ul>
      <button
        class="sc-ewnqHT ctMhSl"
        type="button"
      >
        Load more
      </button>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
